### DISCUSSION TITLE
複数のサーバーで同じテーブルを持つことに対する違和感

### DISCUSSION THREAD

**Discussion by yuyakawai0411** [kakeru-one/web-engineering-roundtable#33] [Q&A] [answered]
## 背景
> トランザクションデータ用サーバそれぞれに、商品マスタテーブルのデータを全てコピーしておくようにする
https://github.com/ochi-sho-private-study/mysql-sandbox/blob/main/docs/resolve_sql_weak_point/db_bottleneck.md

その商品の最後に売上がたった日を知りたい時、トランザクションデータ用サーバBにアクセスするだけで済むため、スピードが早くなりそうですが、複数のサーバーで同じテーブルを持つことに違和感を感じました。
というのも、新しい商品を追加した直後に売上明細にアクセスする時、トランザクションデータ用サーバBに商品マスタのデータがまだ反映されていないということが起きる気がしています。
関連が強い(ex. 参照が頻繁に発生する)テーブルはなるべく同じサーバーで管理した方がいいと思いますが、データの数によってはそうも言ってられない気がするので、ベストプロアクティスは思いついていないです🙏

**Comment by kakeru-one** [comment, marked as answer]
> - マスタ系データの更新は、マスタデータ用のサーバーに対してのみ行い、各トランザクション用サーバに自動的に反映させるようにする。これを実現する手段としては、以下がある。
>   - DBリンクを使ったマテリアライズドビュー
>   - トリガーでコピーする
>   - DBリンクがないMySQLの場合は、レプリケーションする

この辺が対策になっていて、マスタデータ用のサーバーにinsertやupdateが走ったらトランザクション用サーバに同期するようにするようです。

**Comment by kakeru-one** [comment]
UPDATEやINSERT、DELETEが走る時はどのようなロックがかかるのか？
https://www.ibm.com/docs/ja/informix-servers/12.10?topic=locking-locks-placed-insert-update-delete-statements

**Reply by yuyakawai0411** [reply]
排他ロックがかかるのであれば、SELECTもできなくなるため、過去の情報が参照されることもなく、自分の懸念点は解消される

**Reply by kakeru-one** [reply]
ロックの概要
https://techblog.cartaholdings.co.jp/entry/2022/12/14/113000

**Comment by kakeru-one** [comment]
オンラインDDL
https://dev.mysql.com/doc/refman/8.0/ja/innodb-online-ddl-operations.html#:~:text=%E8%A1%A8%C2%A015.18%C2%A0%E3%82%AB%E3%83%A9%E3%83%A0%E6%93%8D%E4%BD%9C%E3%81%AE%E3%82%AA%E3%83%B3%E3%83%A9%E3%82%A4%E3%83%B3%20DDL%20%E3%82%B5%E3%83%9D%E3%83%BC%E3%83%88
